### PR TITLE
Stdenv setup: Process dependencies before dependents

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

This is intended to be reviewed commit-by-commit.

This does 2 things

1. Change the dependency order. Before, we did the current node before children, now we do the children before current node. Were dependencies forming a tree, this would simply be the inverse traversal, but since they in fact form a DAG, it's a bit more subtle. The opposite of the new invariant "dependencies before dependents" wasn't true: dependents weren't always processed first because it was possible that a dependency was already visited via another dependent.

   This is an observable change because the processing we are doing does not commute: arbitrary bash setup hooks certainly don't, and `PATH` concatenation doesn't either. I do think having this invariant is valuable, though, and rely on it to simplify cc-wrapper's setup hook.

2. Cyclic dependencies are now an error.  Yes, they are only possible in absurd cases, but possible nonetheless:
   ```
   └── /nix/store/asdfasdfasdf-foobar/
       ├── foo
       │   └── nix-support
       │       └── propagated-build-inputs = /nix/store/asdfasdfasdf-foobar/bar
       └── bar
           └── nix-support
               └── propagated-build-inputs = /nix/store/asdfasdfasdf-foobar/foo
   ```
   (`foo` and `bar` are cyclic). The old algo would silently work with this; the new one would naively diverge, so I made it error out instead as such a thing should never be intended. But if we agree this shouldn't happen in practice so infinite recursion is a good enough error, I can simplify things

###### Things done

~~Hoping to merge this **Monday 2017-11-13** at the latest :). Thanks!~~

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built stdenv on platform(s)
   - [ ] NixOS -- latest in progress, see nixborg build
   - [ ] macOS -- latest in progress
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---------

CC @bgamari

### CODE REVIEW

**Comment by copumpkin** [comment]
This seems sensible in principle, but can we get a Hydra job showing what the implications might be? Subtly changing behavior (that arguably nobody should rely on the order of, but people probably do) makes me very uncomfortable even if it does make sense.

**Comment by globin** [comment]
@nixborg build

**Comment by globin** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -310,20 +313,54 @@ findInputs() {
...
+    # top of the stack and pop it.
+    [[ "${inProgressStack[${#inProgressStack[@]} - 1]}" = "$pkg" ]]
+    unset "inProgressStack[${#inProgressStack[@]} - 1]"
+
+    # Mark it completely done
+    eval "$var"'+=("$pkg")'
+}
+
+# Add package to the future PATH and run setup hooks
+activatePackage() {
```
Comment (orivej) [inline pkgs/stdenv/generic/setup.sh]: Who calls this function?

File: pkgs/stdenv/generic/setup.sh [hunk 2]
```diff
@@ -359,15 +389,22 @@ if [ -z "${crossConfig:-}" ]; then
              ${propagatedNativeBuildInputs:-} ${propagatedBuildInputs:-}; do
         findInputs "$i" nativePkgs propagated-native-build-inputs
     done
+    for i in "${nativePkgs[@]}"; do
+        activatePackage "$i"
```
Comment (Ericson2314) [inline pkgs/stdenv/generic/setup.sh]: @orivej here

File: pkgs/stdenv/generic/setup.sh [hunk 3]
```diff
@@ -359,15 +389,22 @@ if [ -z "${crossConfig:-}" ]; then
...
     for i in ${buildInputs:-} ${defaultBuildInputs:-} ${propagatedBuildInputs:-}; do
         findInputs "$i" crossPkgs propagated-build-inputs
     done
-
-    declare -a nativePkgs
     for i in ${nativeBuildInputs:-} ${defaultNativeBuildInputs:-} ${propagatedNativeBuildInputs:-}; do
         findInputs "$i" nativePkgs propagated-native-build-inputs
     done
+    for i in "${nativePkgs[@]}"; do
+        activatePackage "$i"
```
Comment (Ericson2314) [inline pkgs/stdenv/generic/setup.sh]: @orivej And here

**Review by orivej — APPROVED** [review]

**Comment by Ericson2314** [comment]
@globin is there any good known-good jobsest to compare the nixborg one to?

**Comment by Ericson2314** [comment]
Oh hmm, this breaks down when different dependencies are in fact the same derivation. I'm going to have to think about that.

**Comment by orivej** [comment]
Could you give an example in nixpkgs?

**Comment by Ericson2314** [comment]
@orivej Nevermind. The already-visited logic *is* based on actual paths, I just mistakenly changed the iteration order. [I did have to make `_PATH` be prepended instead of appended, but that is expected.]

There's still a theoretical issue where "unrelated" deps could clobber each other on the PATH in different ways based on different topological sorts, and our depth-first one might not have the most intuitive behavior in that regard, but I won't worry about this unless it shows up in practice.

I added a `cc-wrapper` commit from my https://github.com/NixOS/nixpkgs/pull/29396 branch that actually leverages this, to make sure everything works alright.

**Review by orivej — APPROVED** [review]
Looks good. https://hydra.mayflower.de/jobset/nixos/pr-31414 did not pick your changes yet… And since AFAIK it does not contribute to the nixos cache, shouldn't this PR go to staging first?

**Comment by globin** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by Ericson2314** [comment]
Ok, the last build failed because of uses of `addToSearchPathWithDelimiter`. I renamed `addToSearchPath*` to `appendToSearchPath` across the board to be explicit and consistent, avoiding the undefined function problem in the process. This does break the the interface, but as those functions aren't really documented anywhere I hope there was no expectation of stability.

Additionally, I changed some `appends*` to `prepends*` to match the new near-opposite order dependencies are accumulated as. I think there's a good chance that fixes the only potential breakage this PR introduces. [The Nixborg packages were not affected by this either way as the first hydra evaluation shows, oh well.]

Still hoping to merge this tomorrow, but I'll understand if this change to the PR means people need more time to review. I have reorganized commits so the history still tells a clear story.

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

**Comment by Ericson2314** [comment]
Hmm the darwin error is seems to be https://github.com/Andersbakken/rtags/issues/937#issuecomment-293459813, a problem with `#include_next` and the incnlude order.  Still investigating the Linux problem.

My last attempt did not work as well as I thought...

**Comment by Ericson2314** [comment]
Ok, I I did two things which should fix the previous problems:

1. Default dependencies are processed first, so they are still lower priority. This should fix the Darwin problem as it will effectively put libc before just about any other library.

2. Setup hooks are (painstakingly!) edited to ensure the current package supersedes any previously-processed one.

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by Ericson2314** [comment]
Rebased on top of now-merged https://github.com/NixOS/nixpkgs/pull/31723, which was part of this. I doubt that will fix this, but if it does, great, and if it doesn't, the build will fail quickly so that not too many resources were wasted.

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-31414

**Comment by FRidh** [comment]
I like the motivation behind this change but because nothing is happening I am closing this.

**Comment by JohnAZoidberg** [comment]
This issue is mentioned in nixpkgs stdenv: https://github.com/nixos/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L1167

**Comment by Ericson2314** [comment]
Yes maybe a GitHub issue rather than open PR canbe made for that.

**Comment by Ericson2314** [comment]
Reopning and marking as draft because I ought to finish it at some point.

**Comment by Ericson2314** [comment]
Should be possible to finish after #132340 is the default and cc-wrapper is simplified.

**Comment by ConnorBaker** [comment]
I think this is still relevant and (time permitting) I'd like to revive this PR.

@Ericson2314 in terms of how this would look on current `master`, any advice? I started by trying to apply the spirit of the changes each commit contains, but I'm not sure what some of the changes should look like post #26805.

For example, what should https://github.com/NixOS/nixpkgs/pull/31414/commits/229eed32411b2d680757df437073a818a080c0ce look like since https://github.com/NixOS/nixpkgs/pull/26805/commits/3a50395ef226259694689da92b2fd636476abc67? That is, I'm not sure what the offsets are for the wrapper and scripts relative to the consumers.

Additionally, I see a lot of interpolation of derivations into store paths in various strings -- is this a concern with respect to splicing? I know that `mkDerivation` includes code to handle retrieving packages from the correct package set when they are in dependency arrays, but I don't think that applies to the various string arguments passed to `mkDerivation` (nor could it, I suppose).

**Comment by Ericson2314** [comment]
> That is, I'm not sure what the offsets are for the wrapper and scripts relative to the consumers.

I think the offsets can stay the same? Or at least if

- the old PR moved things to propagated deps
- `master` had recategorized which type of dep

Do "propagated" per the PR, and the offset per `master`.

**Pull request closed without merging**